data-structures/filecoin/chain.md

```diff
@@ -0,0 +1,1225 @@
+# Filecoin Main Chain Data Structures
+
+Schemas are grouped by their serialized blocks. Other than those types listed in "Basic Types" and "Crypto Types", each grouping of schema types in a code block represents a data structure that is serialized into a single IPLD block with its own Link (CID).
+
+Advanced Data Layouts (ADLs) are shown in their expanded form here, as the data appears on-block. Their logical forms for programmatic purposes are `Map` for the HAMT and `List` for the AMT.
+
+There are some data structures that are repeats of the same forms, primarily the AMT and HAMTs that share the same data types. They are not de-duplicated here for clarity to demonstrate the different purposes of those data structures.
+
+For more information about the IPLD Schema language, see the [specificaiton](https://specs.ipld.io/schemas/).
+
+## Basic Types
+
+```ipldsch
+# Address is a `https://github.com/filecoin-project/go-address` in its binary byte format.
+# It may be used as a hamt key, and will be often visualized or presented to users
+# in its more readable string form.
+type Address bytes
+
+# Go big.Int
+# Prefer presenting to users either as a number or a string view of the decimal number
+# for readability.
+type BigInt bytes
+
+# An indicator of which RPC method on an actor a message should trigger execution of.
+type MethodNum int
+
+# The 'f0...' subset of Addresses used for the actual indexes of actors in a state root.
+type ActorID int
+
+# the height of a block in the chain. Should fit in an Int64
+type ChainEpoch int
+
+type TokenAmount BigInt
+
+type PaddedPieceSize int
+
+type PeerID bytes
+
+type SectorSize int
+
+type SectorNumber int
+
+type PartitionNumber int
+
+type BitField bytes
+
+type StoragePower BigInt
+
+type DataCap StoragePower
+
+type DealID int
+
+type DealWeight BigInt
+
+type Multiaddr bytes
+
+type RegisteredSealProof int
+
+type TransactionID # TxnID
+```
+
+## Crypto Types
+
+```ipldsch
+type Signature union {
+  SignatureBLS 0
+  SignatureSecp256k1 1
+} representation byteprefix
+```
+
+## Genesis Block
+
+```ipldsch
+type Genesis struct {
+  Datetime String
+  Network String
+  Token String
+  TokenAmounts TokenAmounts
+  Message String
+}
+
+type TokenAmounts struct {
+  TotalSupply String
+  Miners String
+  ProtocolLabs {String:String}
+}
+```
+
+## Chain
+
+```ipldsch
+type BlockHeader struct {
+  Miner Address
+  Ticket nullable Ticket
+  ElectionProof nullable ElectionProof
+  BeaconEntries [BeaconEntry]
+  WinPoStProof [PoStProof]
+  Parents [&Any] # TODO: can we &BlockHeader this or is it a union with Genesis?
```

Comment:
Is it worth defining `[&BlockHeader]` as `TipSetKey`?
It's worth checking if the height=0 block explicitly has parents set to the genesis block, or if we can type this stronger, since it would make things nicer if we could.

Reply:
```
# The type hint here, `BlockHeader`, holds true except for the case of the
# Genesis block, which is a different format entirely.
type TipSetKey [&BlockHeader]
```

going with that for now 🤞, I can't see anything obvious in Lotus that tells me a clear answer to this, but I've asked, so if I get an answer that's different from this advice I'll update it.